pkg/server/tls_windows.go

```diff
@@ -0,0 +1,342 @@
+	default:
+		return fmt.Errorf("invalid ClientAuth %q", c.ClientAuth)
+	}
+	// Kick off the refresh handler
+	go cn.startUpdateTimer()
+	l.windowsCertHandler = cn
+	l.tlsConfig = config
+	l.cfg = c
+	return nil
+}
+
+// this is the ASN1 Object Identifier for TemplateID
+var asnTemplateOID = "1.3.6.1.4.1.311.21.7"
+
+type winCertStoreHandler struct {
+	cfg          WindowsCertificateFilter
+	subjectRegEx *regexp.Regexp
+
+	winMut       sync.Mutex
+	serverCert   *x509.Certificate
+	serverSigner crypto.PrivateKey
+	// We have to store the identity to access the signer (its an api call), the client does NOT need the signer
+	serverIdentity certstore.Identity
+	clientRootCA   *x509.Certificate
+
+	stopUpdateTimer chan struct{}
+}
+
+func (c *winCertStoreHandler) startUpdateTimer() {
+	if c.cfg.ServerRefreshInterval == 0 {
+		c.cfg.ServerRefreshInterval = 5 * time.Minute
+	}
+
+	select {
+	case <-c.stopUpdateTimer:
+		if c.serverIdentity != nil {
+			c.serverIdentity.Close()
+		}
+		c.serverCert = nil
+		c.serverSigner = nil
+		return
+	case <-time.After(c.cfg.ServerRefreshInterval):
+		err := c.refreshCerts()
+		if err != nil {
+			return
+		}
+
+	}
+}
+
+// refreshCerts is the main work item in certificate store, responsible for finding the right certificate
+func (c *winCertStoreHandler) refreshCerts() error {
+	c.winMut.Lock()
+	defer c.winMut.Unlock()
+	// Close the server identity if already set
+	if c.serverIdentity != nil {
+		c.serverIdentity.Close()
+	}
+	var serverIdentity certstore.Identity
+	var clientIdentity certstore.Identity
+	var err error
+	// This handles closing all our various handles
+	defer func() {
+		// we have to keep the server identity open if we want to use it, BUT only if an error occurred, else we need it
+		// open to sign
+		if serverIdentity != nil && err != nil {
+			serverIdentity.Close()
+		}
+		// Client identity does need to be open since we dont need to sign
+		if clientIdentity != nil {
+			clientIdentity.Close()
+		}
+	}()
+	serverIdentity, err = c.findServerCertificate()
+	if err != nil {
+		return err
+	}
+	sc, err := serverIdentity.Certificate()
+	if err != nil {
+		return err
+	}
+	signer, err := serverIdentity.Signer()
+	if err != nil {
+		return err
+	}
+	clientIdentity, err = c.findClientCertificate()
+	if err != nil {
+		return err
+	}
+	cc, err := clientIdentity.Certificate()
+	if err != nil {
+		return err
+	}
+	c.serverCert = sc
+	c.serverSigner = signer
+	c.clientRootCA = cc
+	c.serverIdentity = serverIdentity
+	return nil
+}
+
+func (c *winCertStoreHandler) findServerCertificate() (certstore.Identity, error) {
+	return c.findCertificate(c.cfg.ServerSystemStore, c.cfg.ServerStore, c.cfg.ServerIssuerCommonNames, c.cfg.ServerTemplateID, nil, c.getStore)
+}
+func (c *winCertStoreHandler) findClientCertificate() (certstore.Identity, error) {
+	return c.findCertificate(c.cfg.ClientSystemStore, c.cfg.ClientStore, c.cfg.ClientIssuerCommonNames, c.cfg.ClientTemplateID, c.subjectRegEx, c.getStore)
+}
```

Comment:
Is it required to set both server and client certificates? IMO client should be optional (though server must be set if client is set)

Reply:
Refactored so that it's more intelligent about when to throw if client certs are not set.